File: leetcode/longest_zigzag_path_in_a_binary_tree/longest_zigzag_path_in_a_binary_tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class LeftRightMax {
public:
    int left;
    int right;
    int max_num;
};
class Solution {
public:
    int longestZigZag(TreeNode* root) {
        int ans = 0;
        LeftRightMax rightDirection = DFS(root->right);
        ans = max(rightDirection.left + 1, ans);
        ans = max(rightDirection.right, ans);
        ans = max(rightDirection.max_num, ans);

        LeftRightMax leftDirection = DFS(root->left);
        ans = max(leftDirection.right + 1, ans);
        ans = max(leftDirection.left, ans);
        ans = max(leftDirection.max_num, ans);

        return ans;
    }

    LeftRightMax DFS(TreeNode* node){
        if(!node){
            LeftRightMax leafNode;
            leafNode.left=-1;
            leafNode.right=-1;
            leafNode.max_num=-1;
            return leafNode;
        }

        LeftRightMax leftD = DFS(node->left);
        LeftRightMax rightD = DFS(node->right);

        int left = leftD.right + 1;
        int right = rightD.left + 1;
        int max_num = 0;
        max_num=max(max_num,leftD.left);
        max_num=max(max_num,leftD.max_num);
        max_num=max(max_num,rightD.right);
        max_num=max(max_num,rightD.max_num);

        LeftRightMax return_value;
        return_value.left=left;
        return_value.right=right;
        return_value.max_num=max_num;

        return return_value;
    }
};

File: leetcode/longest_zigzag_path_in_a_binary_tree/longest_zigzag_path_in_a_binary_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def DFS(self, node: Optional[TreeNode], isleft: bool):
        if node==None:
            return 0, 0

        left_connect, left_max=self.DFS(node.left, True)
        right_connect, right_max=self.DFS(node.right, False)

        if isleft:
            connect = right_connect + 1
            max_val = max(left_connect, left_max, right_max)
        else:
            connect = left_connect + 1
            max_val = max(left_max, right_connect, right_max)
        
        return connect, max_val


    def longestZigZag(self, root: Optional[TreeNode]) -> int:
        left_connect, left_max=self.DFS(root.left, True)
        right_connect, right_max=self.DFS(root.right, False)

        return max(left_connect, right_connect, left_max, right_max)